source/validate_types.cpp

```diff
@@ -0,0 +1,206 @@
+#include <vector>
+
+using std::find;
+using std::string;
+using std::to_string;
+using std::unordered_set;
+using std::vector;
+
+namespace {
+const vector<vector<SpvOp>>& GetModuleOrder() {
+  // See Section 2.4
+  // clang-format off
+  static const vector<vector<SpvOp>> moduleOrder = {
+    {SpvOpCapability},
+    {SpvOpExtension},
+    {SpvOpExtInstImport},
+    {SpvOpMemoryModel},
+    {SpvOpEntryPoint},
+    {SpvOpExecutionMode},
+    {
+      // first set of debug instructions
+      SpvOpSourceContinued,
+      SpvOpSource,
+      SpvOpSourceExtension,
+      SpvOpString,
+    },
+    {
+      // second set of debug instructions
+      SpvOpName,
+      SpvOpMemberName
+    },
+    {
+      // annotation instructions
+      SpvOpDecorate,
+      SpvOpMemberDecorate,
+      SpvOpGroupDecorate,
+      SpvOpGroupMemberDecorate,
+      SpvOpDecorationGroup
+    },
+    {
+      // All type and constant instructions
+      SpvOpTypeVoid,
+      SpvOpTypeBool,
+      SpvOpTypeInt,
+      SpvOpTypeFloat,
+      SpvOpTypeVector,
+      SpvOpTypeMatrix,
+      SpvOpTypeImage,
+      SpvOpTypeSampler,
+      SpvOpTypeSampledImage,
+      SpvOpTypeArray,
+      SpvOpTypeRuntimeArray,
+      SpvOpTypeStruct,
+      SpvOpTypeOpaque,
+      SpvOpTypePointer,
+      SpvOpTypeFunction,
+      SpvOpTypeEvent,
+      SpvOpTypeDeviceEvent,
+      SpvOpTypeReserveId,
+      SpvOpTypeQueue,
+      SpvOpTypePipe,
+      SpvOpTypeForwardPointer,
+      SpvOpConstantTrue,
+      SpvOpConstantFalse,
+      SpvOpConstant,
+      SpvOpConstantComposite,
+      SpvOpConstantSampler,
+      SpvOpConstantNull,
+      SpvOpSpecConstantTrue,
+      SpvOpSpecConstantFalse,
+      SpvOpSpecConstant,
+      SpvOpSpecConstantComposite,
+      SpvOpSpecConstantOp,
+      SpvOpVariable,
+      SpvOpLine
+    }
+  };
+  // clang-format on
+
+  return moduleOrder;
+}
+}
+
+namespace libspirv {
+
+ValidationState_t::ValidationState_t(spv_diagnostic* diag, uint32_t options)
+    : diagnostic_(diag),
+      instruction_counter_(0),
+      defined_ids_{},
+      unresolved_forward_ids_{},
+      validation_flags_(options),
+      operand_names_{},
+      module_layout_order_stage_(0),
+      current_layout_stage_(ModuleLayoutSection::kModule) {}
+
+spv_result_t ValidationState_t::defineId(uint32_t id) {
+  if (defined_ids_.find(id) == end(defined_ids_)) {
+    defined_ids_.insert(id);
+  } else {
+    return diag(SPV_ERROR_INVALID_ID) << "ID cannot be assigned multiple times";
+  }
+  return SPV_SUCCESS;
+}
+
+spv_result_t ValidationState_t::forwardDeclareId(uint32_t id) {
+  unresolved_forward_ids_.insert(id);
+  return SPV_SUCCESS;
+}
+
+spv_result_t ValidationState_t::removeIfForwardDeclared(uint32_t id) {
+  unresolved_forward_ids_.erase(id);
+  return SPV_SUCCESS;
+}
+
+void ValidationState_t::assignNameToId(uint32_t id, string name) {
+  operand_names_[id] = name;
+}
+
+string ValidationState_t::getIdName(uint32_t id) const {
+  string out = to_string(id);
```

Comment:
Sorry, std::to_string is not available on Android.  I patched your previous CL before merging into master.  Please pull that fix into this commit.